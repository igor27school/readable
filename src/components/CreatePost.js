import React, { Component } from 'react'
import { connect } from 'react-redux'
import serializeForm from 'form-serialize'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import { fetchCategoriesFromServer, createPost } from '../actions/ActionCreators'

/**
* @description The view for creating a new post.
*/
export class CreatePost extends Component {
  static propTypes = {
    hasCategories: PropTypes.bool.isRequired,
    categories: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
      }).isRequired,
    ).isRequired,
    fetchCategoriesFromServer: PropTypes.func.isRequired,
  }
  componentDidMount() {
    const { hasCategories, fetchCategoriesFromServer } = this.props
    if (!hasCategories) {
      fetchCategoriesFromServer()
    }
  }
  handleSubmit = (e) => {
    e.preventDefault()
    const values = serializeForm(e.target, { hash: true })
    // Basic form data validation
    if (values.category === 'none') {
      alert('Please fill out the category!')
      return false
    }
    if (!values.title || values.title.trim() === '') {
      alert('Please include a proper title!')
      return false
    }
    this.props.createPost(values)
    this.props.history.push('/')
  }
  render() {
    const { categories } = this.props
    return (
      <div>
        <Link to="/">Close</Link>
        <form onSubmit={this.handleSubmit}>
          <input type="text" name="title" placeholder="Title"/>
          <input type="text" name="author" placeholder="Author"/>
          <textarea name="body" placeholder="Body"></textarea>
          <select defaultValue="none" name="category">
            <option value="none" disabled>Select category</option>
            {categories.map((category) => (
              <option
                key={category.name}
                value={category.name}
              >{category.name}</option>
            ))}
          </select>
          <button>Publish</button>
        </form>
      </div>

    )
  }
}

function mapStateToProps ({ categories }) {
  return {
    hasCategories: categories.allIds.length > 0,
    categories: categories.allIds.map(category => categories.byId[category]),
  }
}

function mapDispatchToProps (dispatch) {
  return {
    fetchCategoriesFromServer: () => dispatch(fetchCategoriesFromServer()),
    createPost: values => dispatch(createPost(values)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(CreatePost)
