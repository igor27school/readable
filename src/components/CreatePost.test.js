import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { CreatePost } from './CreatePost'

jest.mock('../utils/ServerAPI')

describe('CreatePost tests', () => {
  const setup = () => {
    const props = {
      hasCategories: true,
      categories: [
        {
          name: 'category1',
          path: 'category1',
        },
        {
          name: 'category2',
          path: 'category2',
        },
      ],
      fetchCategoriesFromServer: jest.fn(),
    }
    const renderer = new ShallowRenderer()
    return {
      props,
      renderer
    }
  }

  it('CreatePost matches the snapshot', () => {
    const { renderer, props } = setup()
    renderer.render(<CreatePost {...props}/>)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
})
