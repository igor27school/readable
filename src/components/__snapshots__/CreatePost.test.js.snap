// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePost tests CreatePost matches the snapshot 1`] = `
<div>
  <Link
    replace={false}
    to="/"
  >
    Close
  </Link>
  <form
    onSubmit={[Function]}
  >
    <input
      name="title"
      placeholder="Title"
      type="text"
    />
    <input
      name="author"
      placeholder="Author"
      type="text"
    />
    <textarea
      name="body"
      placeholder="Body"
    />
    <select
      defaultValue="none"
      name="category"
    >
      <option
        disabled={true}
        value="none"
      >
        Select category
      </option>
      <option
        value="category1"
      >
        category1
      </option>
      <option
        value="category2"
      >
        category2
      </option>
    </select>
    <button>
      Publish
    </button>
  </form>
</div>
`;
